recipe-client-addon/lib/PreferenceManagement.jsm

```diff
@@ -0,0 +1,395 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * PreferenceManagement exposes an API for recipes such as Preference Experiments
+ * to handle manipulating default/user branch preferences. Recipes should extend
+ * this module.
+ *
+ * Info on active and past recipes is stored in a JSON file in the profile
+ * folder. Preference observers can also be tracked/instantiated/etc.
+ */
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "Services", "resource://gre/modules/Services.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "CleanupManager", "resource://shield-recipe-client/lib/CleanupManager.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "JSONFile", "resource://gre/modules/JSONFile.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "OS", "resource://gre/modules/osfile.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "LogManager", "resource://shield-recipe-client/lib/LogManager.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "Preferences", "resource://gre/modules/Preferences.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "TelemetryEnvironment", "resource://gre/modules/TelemetryEnvironment.jsm");
+
+this.EXPORTED_SYMBOLS = ["PreferenceManagement"];
+
+const STORE_FILE = "shield-preference-recipes.json";
+
+const PREFERENCE_TYPE_MAP = {
+  boolean: Services.prefs.PREF_BOOL,
+  string: Services.prefs.PREF_STRING,
+  integer: Services.prefs.PREF_INT,
+};
+
+const DefaultPreferences = new Preferences({defaultBranch: true});
+/**
+ * Enum storing Preference modules for each type of preference branch.
+ * @enum {Object}
+ */
+const PreferenceBranchType = {
+  user: Preferences,
+  default: DefaultPreferences,
+};
+
+const log = LogManager.getLogger("preference-management");
+
+this.PreferenceManagement = (namespace) => ({
```

Comment:
I think ES6 classes would make sense, though we'd presumably still want singleton instances for rollouts/experiments.